### PR TITLE
Implement `UnitsFormatter`

### PR DESCRIPTION
**Pull request by younies**

### CODE REVIEW

File: components/experimental/src/dimension/units/options.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /// Format the units with the narrow units format.
+    ///
+    /// The narrow symbol may be ambiguous, so it should be evident from context which
+    /// units is being represented. // TODO??
+    ///
+    /// For example, 1 hour formats as "1 h" in most locales.
+    Narrow,
+}
+
+impl Default for Width {
```
Comment (younies) [inline components/experimental/src/dimension/units/options.rs:46]: TODO: younies, what is the default width for units ?

Comment (younies) [reply components/experimental/src/dimension/units/options.rs:46]: short

File: components/experimental/src/dimension/provider/units.rs [hunk 2]
```diff
@@ -13,6 +13,16 @@ use icu_plurals::PluralCategory;
...
+#[cfg(feature = "compiled_data")]
+/// Baked data
+///
+/// <div class="stab unstable">
+/// 🚧 This code is considered unstable; it may change at any time, in breaking or non-breaking ways,
+/// including in SemVer minor releases. In particular, the `DataProvider` implementations are only
+/// guaranteed to match with this version's `*_unstable` providers. Use with caution.
+/// </div>
+pub use crate::provider::Baked;
+
```
Comment (sffc) [inline components/experimental/src/dimension/provider/units.rs:25]: Issue: Unrelated change

File: components/experimental/src/dimension/units/format.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        let plural_category = self.plural_rules.category_for(self.value);
+        let count = Count::from(plural_category);
+        let unit_pattern = "{0} ".to_owned() + self.unit;
+        let display_name = match self.options.width {
+            Width::Short => self.display_name.short.get(&count),
+            Width::Long => self.display_name.long.get(&count),
+            Width::Narrow => self.display_name.narrow.get(&count),
+        }
+        .unwrap_or({ unit_pattern.as_str() });
```
Comment (sffc) [inline components/experimental/src/dimension/units/format.rs:42]: Suggestion: put the `"{0} ".to_owned() + self.unit` inside `unwrap_or_else` so we don't need to do it each time. Actually, we should just avoid this entirely by making the data struct required to include at least one entry. See example in https://github.com/unicode-org/icu4x/blob/main/documents/design/data_safety.md

File: components/experimental/src/dimension/units/format.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        pattern
+            .interpolate((self.fixed_decimal_formatter.format(self.value),))
+            .write_to(sink)?;
+
+        Ok(())
+    }
+}
+
+#[test]
+fn test_formatted_unit() {
```
Comment (sffc) [inline components/experimental/src/dimension/units/format.rs:56]: Suggestion: Elsewhere in the code base we call these initial smoke tests `test_basic`

```suggestion
fn test_basic() {
```

File: components/experimental/src/dimension/units/formatter.rs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+use icu_provider::DataPayload;
+
+use super::super::provider::units::UnitsDisplayNameV1Marker;
+use super::format::FormattedUnit;
+use super::options::UnitsFormatterOptions;
+use icu_provider::prelude::*;
+
+extern crate alloc;
+
+/// A formatter for monetary values.
```
Comment (sffc) [inline components/experimental/src/dimension/units/formatter.rs:20]:
```suggestion
/// A formatter for measurement unit values.
```

File: components/experimental/src/dimension/units/formatter.rs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    plural_rules: PluralRules,
+}
+
+impl UnitsFormatter {
+    icu_provider::gen_any_buffer_data_constructors!(
+        (locale, unit: &str, options: super::options::UnitsFormatterOptions) -> error: DataError,
+        functions: [
+            try_new: skip,
+            try_new_with_any_provider,
+            try_new_with_buffer_provider,
```
Comment (sffc) [inline components/experimental/src/dimension/units/formatter.rs:50]: Thought: Technically cross-crate constructors should use custom logic, but let's not worry about that at the time being because it's a little annoying

File: components/experimental/src/dimension/units/options.rs (hunk 1 above)
Comment (sffc) [reply components/experimental/src/dimension/units/options.rs:46]: Suggestion: annotate the default variant with `#[default]` and then use `derive(Default)`

File: components/experimental/src/dimension/units/formatter.rs (hunk 6 above)
Comment (younies) [reply components/experimental/src/dimension/units/formatter.rs:50]: sure, I will keep this in mind

File: components/experimental/src/dimension/provider/units.rs (hunk 2 above)
Comment (younies) [reply components/experimental/src/dimension/provider/units.rs]: done, removed

File: components/experimental/src/dimension/units/format.rs (hunk 3 above)
Comment (younies) [reply components/experimental/src/dimension/units/format.rs]: done, thanks for the advice :)

File: provider/bikeshed/src/units/data.rs [hunk 7]
```diff
@@ -15,6 +15,7 @@ use icu::experimental::dimension::provider::units::{
 use icu::locale::LanguageIdentifier;
 use icu_provider::prelude::*;
 use icu_provider::DataMarkerAttributes;
+use icu_provider::DataProvider;
```
Comment (sffc) [inline provider/bikeshed/src/units/data.rs:18]:
```suggestion
```

File: components/experimental/src/dimension/units/options.rs (hunk 1 above)
Comment (sffc) [reply components/experimental/src/dimension/units/options.rs:46]: ^ this is not done

File: components/experimental/src/dimension/units/format.rs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        let mut unit_pattern = None;
+        let display_name = match self.options.width {
+            Width::Short => self.display_name.short.get(&count),
+            Width::Long => self.display_name.long.get(&count),
+            Width::Narrow => self.display_name.narrow.get(&count),
+        }
+        .unwrap_or_else(|| unit_pattern.insert("{0} ".to_owned() + self.unit));
+
+        let pattern =
+            SinglePlaceholderPattern::from_str(display_name).map_err(|_| core::fmt::Error)?;
```
Comment (sffc) [inline components/experimental/src/dimension/units/format.rs:45]: Nit: Please add a TODO

Comment (younies) [reply components/experimental/src/dimension/units/format.rs:46]: done
